### ISSUE TITLE
How can I use callbacks to log to MLflow when using `model.tune()`?

### ISSUE THREAD

**Issue by mesllo-bc** [ultralytics/ultralytics#24699]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

1. I noticed that my callbacks only trigger when I run either `model.train()` or `model.val()`. These then log stuff to MLflow, but when I use `model.tune()` nothing seems to get logged (not even the training stuff even though in the code I see that the `tune` function calls the training anyway. How can I ensure that these are called in the same manner as when I train or validate when tuning?

2. How does tuning work exactly? If I have 10 epochs and 20 tuning iterations, does that mean that the total run would be 200 epochs (10 epochs per tuning iteration)? If so, are there any guidelines on how to adjust these numbers accordingly?

Thank you :)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mesllo-bc, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your question routed quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training or tuning ❓ Question, please provide as much information as possible, including your callback setup, how you are calling `model.tune()`, dataset details, relevant training or tuning logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips). A small reproducible snippet showing how MLflow logging is configured would be especially helpful 🙏

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help the team review this faster, please also share:
- your `ultralytics` version
- your Python, PyTorch, and OS versions
- whether this behavior happens in a clean environment after upgrading
- the exact callbacks you expect to fire during tuning
- any console output or logs showing the difference between `model.train()` and `model.tune()` logging behavior 🧪

**Comment by raimbekovm** [comment]
the reason your callbacks don't fire is that tuning trains each iteration in a separate subprocess (this avoids dataloader hangs across iterations), and callbacks you attach with `model.add_callback(...)` live in your parent python process, so they can't cross into the subprocess. the tuner also doesn't replay your in-process callbacks around each training, which is why nothing logs even though `tune` calls training internally.

the supported way to get per-iteration MLflow logging is the [built-in MLflow integration](https://docs.ultralytics.com/integrations/mlflow/) rather than a custom callback. it's configured through settings and environment variables, and the subprocess inherits both, so every tuning iteration logs as its own MLflow run automatically:

```bash
yolo settings mlflow=True
export MLFLOW_TRACKING_URI=http://127.0.0.1:5000
export MLFLOW_EXPERIMENT_NAME=my_tuning
```

then run `model.tune(...)` as usual. each iteration's training registers the MLflow callbacks on its own trainer and logs params, metrics, and artifacts as a distinct run under that experiment. a custom `add_callback` function can't be used here because it only exists in memory in the parent process, so anything you need during tuning has to come from the env/settings-driven integration, not an in-process callback.

a custom `on_fit_epoch_end` callback fires normally under `model.train()` but never fires when the same callback is added before `model.tune(...)`, while the iteration still trains, so it's the subprocess split, not a registration problem on your side. with the built-in integration enabled, a 3-iteration tune produced 3 separate MLflow runs, each with the training params and metrics logged.

`iterations` is the number of hyperparameter sets the genetic algorithm tries, and each iteration runs one full training of `epochs` epochs. so 10 epochs with 20 iterations is 20 separate trainings of 10 epochs each, i.e. 200 epochs of compute total.

for picking the numbers, the main guideline is to tune under conditions close to your real training (similar dataset, epochs, and augmentations), because hyperparameters found on very short runs rarely transfer to a long final run. `iterations` matters more than it looks: the mutation strength decays over roughly the first 300 iterations, so the search only explores meaningfully when you give it enough iterations. keep `epochs` long enough that the fitness score is a stable signal, but short enough that you can still afford many iterations within your time budget.

the [callbacks guide](https://docs.ultralytics.com/usage/callbacks/) covers the in-process callback API, and the [settings table](https://docs.ultralytics.com/quickstart/) lists the `mlflow` flag. results, including `best_hyperparameters.yaml` and the fitness plot, land under `runs/{task}/tune/`.

**Comment by mesllo-bc** [comment]
I was thinking of trying with 10 epochs over 20 iterations. The mAP signal is still rising at this point but I'm hoping that the tuning can give me an indication of a higher potential even at that point. 300 iterations sounds quite impossible to me given that even at just 10 epochs per iteration that means I have to wait a total of 3000 epochs, correct? If it takes roughly 30 minutes per epoch then I have to wait ~60 days until the true end of the experiment. Is my assumption correct here, or is there actually some heuristic that stops the experiment earlier always?

**Comment by glenn-jocher** [comment]
Yes, your estimate is the correct worst-case budget: `20` iterations with `epochs=10` means up to `200` total training epochs, and `300` iterations would be up to `3000`; the current [tuner implementation](https://docs.ultralytics.com/reference/engine/tuner) does not add a separate tuner-level early-stop heuristic for the search itself, so it will keep going until `iterations` is reached unless an individual training run ends early from normal training behavior such as `patience`, fails, or you stop it manually.

**Comment by raimbekovm** [comment]
the tuner is anytime, so you don't have to commit to the full budget up front. it writes `best_hyperparameters.yaml` and the fitness plot after every iteration, so you can watch the curve, stop with ctrl-c whenever it plateaus, and resume later with `resume=True` (passing the same `data`, `epochs`, `iterations`, and `space`). 20 iterations is a fine starting point for a directional signal and you can extend it by resuming rather than restarting.

if you want to fit more iterations into the same wall-clock, lower the per-iteration cost rather than the count: `fraction<1.0` to tune on a subset of the data, a smaller `imgsz`, and `cache=ram` or `disk`, plus `plots=False, save=False`. the trade-off noted in the [tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/) is that hyperparameters found on short or reduced runs can bias toward fast convergence and may not fully transfer, so treat the result as directional and re-validate the top config at your full training settings before committing.

**Comment by mesllo-bc** [comment]
Perfect. I'm already using `patience` in my experiments so I'm hoping early-stopping could also help keep CPU/GPU usage at a minimum. Thanks for your insight everyone!

**Comment by glenn-jocher** [comment]
Sounds good — `patience` applies within each training run during `model.tune()`, so it can reduce compute when a trial plateaus early; if you hit any tuning-specific MLflow issue later, please open a new issue with `yolo checks` output and the full traceback and we’ll take a look.
